[fix] clone should preserve the type of attribute

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29269 [fix] clone should preserve the type of attribute**

Summary:
Hit this bug when I have an attribute of type `Optional[Tensor]` which
is initialized to None and reassigned later to some tensor.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18364338](https://our.internmc.facebook.com/intern/diff/D18364338)

jerryzh168 committed:
[fix] clone should preserve the type of attribute

Summary:
Hit this bug when I have an attribute of type `Optional[Tensor]` which
is initialized to None and reassigned later to some tensor.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito reviewed (approved):
Thanks -- can you wait until https://github.com/pytorch/pytorch/pull/28828 lands? It conflicts with this patch and has been getting delayed by merge conflicts that past few days.

jerryzh168 commented:
> Thanks -- can you wait until #28828 lands? It conflicts with this patch and has been getting delayed by merge conflicts that past few days.

Sure

suo commented:
Thanks for the fix. Can we add tests for `copy()` as well? This was broken for some time before anyone realized because there are no tests. Thanks!

jerryzh168 committed:
Update on "[fix] clone should preserve the type of attribute"


Summary:
Hit this bug when I have an attribute of type `Optional[Tensor]` which
is initialized to None and reassigned later to some tensor.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fix] clone should preserve the type of attribute"


Summary:
Hit this bug when I have an attribute of type `Optional[Tensor]` which
is initialized to None and reassigned later to some tensor.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fix] clone should preserve the type of attribute"


Summary:
Hit this bug when I have an attribute of type `Optional[Tensor]` which
is initialized to None and reassigned later to some tensor.

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18364338](https://our.internmc.facebook.com/intern/diff/D18364338)

[ghstack-poisoned]
